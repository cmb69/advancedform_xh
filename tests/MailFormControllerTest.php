<?php

namespace Advancedform;

use Advancedform\Infra\CaptchaWrapper;
use Advancedform\Infra\HooksWrapper;
use Advancedform\Infra\Logger;
use Advancedform\PHPMailer\PHPMailer;
use ApprovalTests\Approvals;
use org\bovigo\vfs\vfsStream;
use PHPUnit\Framework\MockObject\MockObject;
use PHPUnit\Framework\MockObject\Stub;
use PHPUnit\Framework\TestCase;
use Plib\FakeRequest;
use Plib\View;

class MailFormControllerTest extends TestCase
{
    /** @var FormGateway */
    private $formGateway;

    /** @var FieldRenderer */
    private $fieldRenderer;

    /** @var Validator */
    private $validator;

    /** @var CaptchaWrapper&Stub */
    private $captchaWrapper;

    /** @var HooksWrapper&MockObject */
    private $hooksWrapper;

    /** @var PHPMailer&MockObject */
    private $mailer;

    /** @var MailService */
    private $mailService;

    /** @var Logger&Stub */
    private $logger;

    public function setUp(): void
    {
        vfsStream::setUp("root");
        mkdir(vfsStream::url("root/data"));
        copy("./data/forms.json", vfsStream::url("root/data/forms.json"));
        mkdir(vfsStream::url("root/css"));
        copy("./css/stylesheet.css", vfsStream::url("root/css/stylesheet.css"));
        $this->formGateway = new FormGateway(vfsStream::url("root/data/"));
        $this->hooksWrapper = $this->createMock(HooksWrapper::class);
        $this->hooksWrapper->method("validField")->willReturn(true);
        $this->hooksWrapper->method("mail")->willReturn(true);
        $this->fieldRenderer = new FieldRenderer("Contact", $this->hooksWrapper);
        $this->captchaWrapper = $this->createStub(CaptchaWrapper::class);
        $lang = XH_includeVar("./languages/en.php", "plugin_tx")["advancedform"];
        $this->validator = new Validator(
            XH_includeVar("./config/config.php", "plugin_cf")["advancedform"],
            $lang,
            $this->captchaWrapper,
            $this->hooksWrapper
        );
        $this->mailer = $this->getMockBuilder(PHPMailer::class)
            ->onlyMethods(["Send"])
            ->getMock();
        $this->mailService = new MailService(
            vfsStream::url("root/data/"),
            vfsStream::url("root/"),
            $lang,
            $this->mailer,
            $this->hooksWrapper
        );
        $this->logger = $this->createStub(Logger::class);
    }

    private function sut(?array $config = null): MailFormController
    {
        return new MailFormController(
            $this->formGateway,
            $this->fieldRenderer,
            $this->validator,
            $this->captchaWrapper,
            $this->hooksWrapper,
            $config ?? $this->config(),
            $this->mailService,
            $this->logger,
            new View("./templates/", XH_includeVar("./languages/en.php", "plugin_tx")["advancedform"])
        );
    }

    public function testRendersMailForm(): void
    {
        $this->captchaWrapper->method("include")->willReturn(true);
        Approvals::verifyHtml($this->sut()->main("Contact", new FakeRequest())->output());
    }

    public function testReportsMissingForm(): void
    {
        $response = $this->sut()->main("NoSuchForm", new FakeRequest());
        $this->assertStringContainsString("A form with the name 'NoSuchForm' does not exist!", $response->output());
    }

    public function testReportsMissingCaptcha(): void
    {
        $this->captchaWrapper->method("include")->willReturn(false);
        $response = $this->sut()->main("Contact", new FakeRequest());
        $this->assertStringContainsString("Could not load CAPTCHA!", $response->output());
    }

    public function testInvalidFormSubmissionRendersValidationErrors(): void
    {
        $_POST = [
            "advfrm" => "Contact",
            "advfrm-E_Mail" => "john",
        ];
        $this->captchaWrapper->method("include")->willReturn(true);
        $response = $this->sut()->main("Contact", new FakeRequest(["post" => ["advfrm" => "Contact"]]));
        $this->assertStringContainsString(
            "Field 'E-Mail' doesn't contain a valid e-mail address!",
            $response->output()
        );
    }

    public function testFailureToSendMailIsReported(): void
    {
        global $e;
        $_SERVER["SERVER_NAME"] = "example.com";
        $_POST = $this->post();
        $e = "";
        $this->captchaWrapper->method("include")->willReturn(true);
        $this->captchaWrapper->method("check")->willReturn(true);
        $this->mailer->method("Send")->willReturn(false);
        $this->sut()->main("Contact", new FakeRequest(["post" => ["advfrm" => "Contact"]]));
        $this->assertSame("<li>The e-mail could not be sent!</li>\n", $e);
    }

    public function testSuccessfulSubmissionsSendsMail(): void
    {
        $_SERVER["SERVER_NAME"] = "example.com";
        $_POST = $this->post();
        $this->captchaWrapper->method("include")->willReturn(true);
        $this->captchaWrapper->method("check")->willReturn(true);
        $this->mailer->method("Send")->willReturn(true);
        $this->sut()->main("Contact", new FakeRequest(["post" => ["advfrm" => "Contact"]]));
        Approvals::verifyHtml($this->mailer->Body);
    }

    public function testShowsFormIfConfirmationMailCouldNotBeSent(): void
    {
        $_SERVER["SERVER_NAME"] = "example.com";
        $_POST = $this->post();
        $config = $this->config();
        $config["mail_confirmation"] = true;
        $this->captchaWrapper->method("include")->willReturn(true);
        $this->captchaWrapper->method("check")->willReturn(true);
        $this->mailer->method("Send")->willReturnOnConsecutiveCalls(true, false);
        $this->hooksWrapper->method("thanksPage")->willReturn("ThankYou");
        $response = $this->sut($config)->main("Contact", new FakeRequest(["post" => ["advfrm" => "Contact"]]));
        $this->assertStringContainsString("<div class=\"advfrm-mailform\">", $response->output());
    }

    public function testRedirectsToThanksPage(): void
    {
        $_SERVER["SERVER_NAME"] = "example.com";
        $_POST = $this->post();
        $config = $this->config();
        $config["mail_confirmation"] = true;
        $this->captchaWrapper->method("include")->willReturn(true);
        $this->captchaWrapper->method("check")->willReturn(true);
        $this->mailer->method("Send")->willReturn(true);
        $this->hooksWrapper->method("thanksPage")->willReturn("ThankYou");
        $response = $this->sut($config)->main("Contact", new FakeRequest(["post" => ["advfrm" => "Contact"]]));
        $this->assertSame("http://example.com/?ThankYou", $response->location());
    }

    public function testSuccessfulSubmissionsRendersMailInfo(): void
    {
        $_SERVER["SERVER_NAME"] = "example.com";
        $_POST = $this->post();
        $this->captchaWrapper->method("include")->willReturn(true);
        $this->captchaWrapper->method("check")->willReturn(true);
        $this->mailer->expects($this->once())->method("Send")->willReturn(true);
        $response = $this->sut()->main("Contact", new FakeRequest(["post" => ["advfrm" => "Contact"]]));
        Approvals::verifyHtml($response->output());
    }

    private function post(): array
    {
        return [
            "advfrm" => "Contact",
            "advfrm-Name" => "John Doe",
            "advfrm-E_Mail" => "john@example.com",
            "advfrm-Comment" => "a comment",
        ];
    }

    private function config(): array
    {
        return XH_includeVar("./config/config.php", "plugin_cf")["advancedform"];
    }
}
