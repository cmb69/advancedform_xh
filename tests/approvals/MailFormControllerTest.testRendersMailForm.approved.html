<!-- advancedform/mail-form -->
<div class="advfrm-mailform">
  <form name="Contact" action="/" method="post" enctype="multipart/form-data" accept-charset="UTF-8">
    <input type="hidden" name="advfrm" value="Contact">
    <div class="required">Please fill in all required fields&nbsp;<span style="color:red">*</span>!</div>
    <!-- advancedform/mail-form-default-view -->
<div style="overflow:auto">
  <table>
    <tr >
      <td class="label">
        <label for="advfrm-Contact-Name">
          Name&nbsp;<span style="color:red">*</span>        </label>
      </td>
      <td class="field"><input type="text" id="advfrm-Contact-Name" name="advfrm-Name" value=""></td>
    </tr>
    <tr >
      <td class="label">
        <label for="advfrm-Contact-E_Mail">
          E-Mail&nbsp;<span style="color:red">*</span>        </label>
      </td>
      <td class="field"><input type="text" id="advfrm-Contact-E_Mail" name="advfrm-E_Mail" value=""></td>
    </tr>
    <tr >
      <td class="label">
        <label for="advfrm-Contact-Phone">
          Phone        </label>
      </td>
      <td class="field"><input type="text" id="advfrm-Contact-Phone" name="advfrm-Phone" value=""></td>
    </tr>
    <tr >
      <td class="label">
        <label for="advfrm-Contact-Comment">
          Comment&nbsp;<span style="color:red">*</span>        </label>
      </td>
      <td class="field"><textarea id="advfrm-Contact-Comment" name="advfrm-Comment" cols="40" rows="4"></textarea></td>
    </tr>
    <tr >
      <td class="label">
                </td>
      <td class="field"><strong>You <em>can</em> attach a file, if you want.</strong></td>
    </tr>
    <tr >
      <td class="label">
        <label for="advfrm-Contact-Attachment">
          Attachment        </label>
      </td>
      <td class="field"><input type="hidden" name="MAX_FILE_SIZE" value="100000"><input type="file" id="advfrm-Contact-Attachment" name="advfrm-Attachment" accept=".jpeg,.jpg,.png,.zip"></td>
    </tr>
    <tr >
      <td class="label">
                </td>
      <td class="field">If you send a password protected zip file, please specify the password here:</td>
    </tr>
    <tr >
      <td class="label">
        <label for="advfrm-Contact-Password">
          Password        </label>
      </td>
      <td class="field"><input type="password" id="advfrm-Contact-Password" name="advfrm-Password" value=""></td>
    </tr>
    <tr >
      <td class="label">
        <label for="advfrm-Contact-Date">
          Date        </label>
      </td>
      <td class="field"><input type="date" id="advfrm-Contact-Date" name="advfrm-Date" value="" placeholder="2019-03-24"></td>
    </tr>
    <tr >
      <td class="label">
                </td>
      <td class="field"></td>
    </tr>
    <tr  class="hidden">
      <td></td>
      <td class="field"><input type="hidden" id="advfrm-Contact-Hidden" name="advfrm-Hidden" value="This is a hidden field!"></td>
    </tr>
  </table>
</div>
        <div class="buttons">
      <input type="submit" class="submit" value="Send mail">
      &nbsp;
      <input type="reset" class="submit" value="Reset form">
    </div>
  </form>
</div>
